### PR TITLE
(PUP-9027) Deprecate CA file settings

### PR DESCRIPTION
**Pull request by Magisus**
In Puppet 6, as part of splitting the CA directory out of the SSL
directory, we are also moving the settings related to CA file locations
in puppetserver's config file. Any settings in Puppet's config file will
be ignored starting then.

### CODE REVIEW

**Comment by Magisus** [comment]
So there are additional CA-related settings in Puppet, this just deprecates the ones that have already been moved by Amy's PR: https://github.com/puppetlabs/puppetserver/pull/1728/files#diff-feeac80211305477061ba2a798669e8dR148 It will take some additional work to see where and how the other related settings are used, but I think that work should be done after that PR is merged.

**Comment by Magisus** [comment]
jenkins please test this servertests

File: lib/puppet/defaults.rb [hunk 1]
```diff
@@ -985,31 +985,43 @@ def self.valid_file_checksum_types
...
     :cacert => {
       :default => "$cadir/ca_crt.pem",
       :type => :file,
       :owner => "service",
       :group => "service",
       :mode => "0644",
-      :desc => "The CA certificate."
+      :desc => "The CA certificate.",
+      :hook => proc { |value|
+        Puppet.deprecation_warning(_("The 'cacert' setting is deprecated. Starting in Puppet 6, this setting will be ignored in favor of the analagous setting in Puppet Server's config."))
```
Comment (justinstoller) [inline lib/puppet/defaults.rb]: Is this a thing where all of these become one translatable string if we pull out the setting name and interpolate it? (that's the case in clojure, not sure about ruby)

File: lib/puppet/defaults.rb [hunk 2]
```diff
@@ -985,31 +985,43 @@ def self.valid_file_checksum_types
       :owner => "service",
       :group => "service",
       :mode => "0755",
-      :desc => "The root directory for the certificate authority."
```
Comment (justinstoller) [inline lib/puppet/defaults.rb]: I'm really unclear here what the best practice is. I see that other deprecated settings have their description updated. Some issue a deprecation warning in a hook and a couple have a key called `:deprecated` with a value of `:completely` (See `trusted_server_facts`).

File: lib/puppet/defaults.rb (hunk 1 above)
Comment (Magisus) [reply lib/puppet/defaults.rb]: Yeah I think so. And I think I just thought of a way to do it.

File: lib/puppet/defaults.rb (hunk 2 above)
Comment (Magisus) [reply lib/puppet/defaults.rb]: Yeah I was wondering the same :/ I think for sure updating the description would be good (I put this up mostly to run the tests on it, was planning to add that). But I have no idea what the `:deprecated` key is used for. I'll look into it.

Comment (Magisus) [reply lib/puppet/defaults.rb]: I got the warning in first because that seemed the most important to me for surfacing it to people

Comment (Magisus) [reply lib/puppet/defaults.rb]: Descriptions updated.

Comment (joshcooper) [reply lib/puppet/defaults.rb]: Using :deprecated and :completely is preferred. Issuing deprecation warnings in hooks predates the new way. Or if only a certain value is deprecated.

Comment (Magisus) [reply lib/puppet/defaults.rb]: Good to know. I need to update this to get all the other CA settings anyway. Thanks Josh!

File: lib/puppet/defaults.rb [hunk 3]
```diff
@@ -1072,6 +1075,8 @@ def self.valid_file_checksum_types
       :type => :autosign,
       :desc => "Whether (and how) to autosign certificate requests. This setting
         is only relevant on a puppet master acting as a certificate authority (CA).
+        This setting is also deprecated and will be replaced by one in Puppet Serve's
```
Comment (justinstoller) [inline lib/puppet/defaults.rb]: missing an 'r' in Puppet Serve's.

**Comment by justinstoller** [comment]
We should also at least update the description of dns_alt_names even if we don't deprecate it (should we deprecate it as well?)

**Comment by Magisus** [comment]
I don't think we should deprecate it, because I think it's still used when creating CSRs to submit to the master, especially if we're going to support signing those with a remote CA. (Thinking of potential easier workflows for non-CA master initialization)

As far as updating the description, what were you thinking? Just that there's a different setting in Puppet Server that controls the SANs of the CA itself?

**Comment by Magisus** [comment]
So currently we get the master's altnames out of that setting in puppet: https://github.com/puppetlabs/puppetserver/blob/master/src/clj/puppetlabs/puppetserver/certificate_authority.clj#L623 Do we want to keep it that way and have the setting in puppetserver apply only to the CA? Or make a new separate setting for the master? Not sure what makes more sense.

**Comment by justinstoller** [comment]
I think it makes sense to have certs for the different roles: CA, Master, and Agent, but I don't think we're going to get there and I don't know how to decide how to only partially split them.

**Comment by Magisus** [comment]
Well right now puppetserver's CA code looks at the puppet setting to select SANs for the "master cert" it generates. Do we want to keep using the puppet setting that, or move it to puppetserver?

(Side note, I just realized my alt names PR is wrong. We don't currently add any alt-names to the CA cert, and probably shouldn't to the intermediate CA cert either.)

**Comment by justinstoller** [comment]
Ugh, that sucks about the gem PR.

If that setting is *only* for the master then... we should move it, but not as part of the CA work.

**Comment by Magisus** [comment]
It's not only for the master, you could use it to specify alt names for agents as well, assuming you enabled signing CSRs with SANs. We just only add defaults if we are the CA (in the old world, this would happen when submitting a CSR for the master's cert).

**Comment by Magisus** [comment]
In that case, I think the comment has been updated correctly, just referencing the deprecations of the mentioned CLI tools.

**Comment by justinstoller** [comment]
Definitely appreciate the review, thanks Josh!

**Pull request merged**